runtime/lua/vim/lsp/document_color.lua

```diff
@@ -0,0 +1,296 @@
+local api = vim.api
+local lsp = vim.lsp
+local util = lsp.util
+local ms = lsp.protocol.Methods
+
+local document_color_ns = api.nvim_create_namespace('nvim.lsp.document_color')
+local document_color_augroup = api.nvim_create_augroup('nvim.lsp.document_color', {})
+
+local M = {}
+
+--- @alias vim.lsp.document_color.HighlightInfo { hl_group: string, range: Range4 }
+
+--- @class (private) vim.lsp.document_color.GlobalState
+--- @field enabled boolean
+local global_state = { enabled = false }
+
+--- @class (private) vim.lsp.document_color.BufState : vim.lsp.document_color.GlobalState
+--- @field buf_version? integer Buffer version for which the color ranges correspond to.
+--- @field applied_version? integer Last buffer version for which we applied color ranges.
+--- @field hl_infos? table<integer, vim.lsp.document_color.HighlightInfo[]> client_id -> processed color highlights
+
+--- @type table<integer, vim.lsp.document_color.BufState>
+local bufstates = vim.defaulttable(function(_)
+  return setmetatable({}, {
+    __index = global_state,
+    __newindex = function(state, key, value)
+      if global_state[key] == value then
+        rawset(state, key, nil)
+      else
+        rawset(state, key, value)
+      end
+    end,
+  })
+end)
+
+--- @param r integer
+--- @param g integer
+--- @param b integer
+local function get_contrast_color(r, g, b)
+  local luminance = 0.298912 * r + 0.586611 * g + 0.114478 * b
+  local is_bright_color = luminance > 127
+  return is_bright_color and '#000000' or '#ffffff'
+end
+
+--- Cache of the highlight groups that we've already created.
+--- @type table<string, true>
+local color_cache = {}
+
+--- Gets or creates the highlight group for the given LSP color information.
+---
+--- @param color lsp.Color
+--- @return string
+local function get_hl_group(color)
+  -- The RGB values in lsp.Color are in the [0-1] range, but we want them to be in the [0-255] range instead.
+  --- @param n number
+  color = vim.tbl_map(function(n)
+    return math.floor((n * 255) + 0.5)
+  end, color)
+  local r, g, b = color.red, color.green, color.blue --- @type integer, integer, integer
+
+  local hex = ('#%02x%02x%02x'):format(r, g, b)
+  local hl_name = 'LspDocumentColor' .. hex:sub(2)
+
+  if not color_cache[hl_name] then
+    api.nvim_set_hl(0, hl_name, { bg = hex, fg = get_contrast_color(r, g, b) })
```

Comment:
Quoting @gpanders and @clason from our Matrix conversation:
> thinking ahead, we will probably want to go ahead and add some kind of function option so that users can make whatever UI they want (like @ribru17 did)
> perhaps pass the buffer number, position of the text (start row+col and end row+col), and the "resolved" hex color, and then the function creates whatever UI element it wants. we could still add some "common" ones as defaults too of course (similar model to what we do for diagnostics)

> I think `background`, `foreground`, and `virtual swatch` would be reasonable (low complexity) ootb options

Reply:
Okay I've added this :)